12.1.1:
The threads try to add at exactly the same time, and incrementing the size at the same time. 
Both threads read a value of 5, increment it to 6 and write it to the set. 
Resulting in a size of 6 when it actually is 7.

12.1.2:
Same as before. Most likely the internal counter that has an error.

12.1.3:
Adding "synchronized" to the "add" and "remove" operations ensures that only one thread can execute at the same time.

12.1.4:
There were no failed tests.

12.1.5:
Not necessarily. It could just be a bad implementation.

12.1.6:
No amount of tests can prove that a collection is thread-safe. Something unexpected can always happen, regardless of the thread-safety of the collection.